PCM rework

huang714 opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
closes #1646 
close #1645 

##### What are the significant changes in functionality due to this change request?
Adding PCM validation capability for steady state problems

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

Jimmy-INL reviewed (changes requested):
Dongli, the XML file has not been pushed yet.

Jimmy-INL commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -137,6 +137,7 @@ def _evaluate(self, datasets, **kwargs):
       msr_mean = 0.0
       msr_std = 0.01*np.std(yapp_std)
       msr_numSmpl = 1000
+      np.random.seed(0)
```
We do not use the NumPy random seed, Please use the one in the utils.

huang714 commented:
Sorry, I cannot find the module to set random seed using utils. Could you please let me know how? Thank you.
________________________________
From: Jimmy-INL ***@***.***>
Sent: Wednesday, December 1, 2021 11:36 PM
To: idaholab/raven ***@***.***>
Cc: Dongli Huang ***@***.***>; Author ***@***.***>
Subject: Re: [idaholab/raven] PCM rework (PR #1726)


@Jimmy-INL requested changes on this pull request.

________________________________

In framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py<https://github.com/idaholab/raven/pull/1726#discussion_r760756176>:

> @@ -137,6 +137,7 @@ def _evaluate(self, datasets, **kwargs):
       msr_mean = 0.0
       msr_std = 0.01*np.std(yapp_std)
       msr_numSmpl = 1000
+      np.random.seed(0)


We do not use the NumPy random seed, Please use the one in the utils.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/raven/pull/1726#pullrequestreview-821041466>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/APV5HEDPCBVPGFCDXMGFYGLUO3ZTVANCNFSM5JCWXCPQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      # Standardization
+      yexp = np.array(featData[0]).reshape(len(featData[0]))
+      yapp = np.array(targData[0]).reshape(len(targData[0]))
+      yexp_ref = np.mean(yexp)
+      yapp_ref = np.mean(yapp)
+      yexp_std = (yexp-yexp_ref)/yexp_ref
+      yapp_std = (yapp-yapp_ref)/yapp_ref
```
Could you rename the variables, because underscore is not allowed in RAVEN code? For example, change 'yexp_std' to 'yexpStd' or 'yexpSTD'

Jimmy-INL replied:
@huang714 , I have addressed all the comments, converted to camelBack, and used randomUtils instead of np.random. This will cause the folded file to change and hence the test is likely to fail. Please doublecheck if the results are correct and then regold again.

huang714 replied:
All variable names with underscore are modified to camelBack format. Golding of the test case is also updated.

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/907364/) on 6bbf96d : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/915871/) on 07152d6 : invalidated by @Jimmy-INL

wangcj05 commented on framework/Metrics/metrics/MetricUtilities.py:
```diff
@@ -227,3 +229,15 @@ def _getPDFCommonArea(data1, data2):
+  """
+  Gets the uncertainty (in form of standard deviation) reduction from data1 to data2.
+  The closer to 1.0 of the reduction fraction, the better representative the experiment is.
+  @ In, data1, data from the prior distribution
+  @ In, data2, data from the posterior distribution
```
please provide the type for 'data1' and 'data2'

huang714 replied:
'a set of data' for both

huang714 replied:
Currently this function is not used

wangcj05 commented on framework/Metrics/metrics/MetricUtilities.py:
```diff
@@ -139,7 +140,8 @@ def _getPDFandCDFfromWeightedData(data, weights, numBins, uniformBins, interpola
   mean = np.average(data, weights = weights)
-  dataStats = {"mean":mean,"minBinSize":minBinSize,"low":low,"high":high}
+  std = np.std(data)
+  dataStats = {"mean":mean,"std":std,"minBinSize":minBinSize,"low":low,"high":high}
```
for consistency, please use "stdev" instead of "std" for the key, or update function "_convertToCommonFormat" to use "std" instead of "stdev".

huang714 replied:
Changed to "stdev".

huang714 replied:
I am not sure whether it is appropriate to use np.std here, as this function seems to work with weighted data.

wangcj05 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,53 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Created on 2017 September 12
+
+@author: Joshua Cogliati
```
please update this lines

huang714 replied:
Updated

wangcj05 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,53 @@
+
+class STDReduction(MetricInterface):
+  """
+    Metric to compare two datasets using the PDF Common Area.
```
please update this docstring, it is not for "PDF Common Area"

huang714 replied:
Updated.

wangcj05 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,53 @@
+      @ Out, NOne
+    """
+    super().__init__()
+    # If True the metric needs to be able to handle (value,probability) where value and probability are lists
+    #self.acceptsProbability = True
+    # If True the metric needs to be able to handle a passed in Distribution
+    #self.acceptsDistribution = True
```
Do you need to keep these lines?

huang714 replied:
I'd like to keep it first, want to make it applicable to probability and value pairs

wangcj05 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,53 @@
+  def run(self, x, y, weights=None, axis=0, **kwargs):
+    """
+      This method computes difference between two points x and y based on given metric
+      @ In, x, instance of Distributions.Distribution, tuple or list, array containing data of x,
+        or given distribution.
+      @ In, y, instance of Distributions.Distribution, tuple or list, array containing data of y,
+        or given distribution.
```
please update the docstring if this function can not handle distribution.

huang714 replied:
Updated.

wangcj05 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,53 @@
+      @ In, kwargs, dict, dictionary of parameters characteristic of each metric
+      @ Out, value, float, metric result, PDF common area
+    """
+    ##TODO: This can be generalized to handle a Distribution
```
Do you want to extend this method to handle a Distribution?

huang714 replied:
I don't think it can handle a Distribution. Deleted.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+  @author: dhuang
+
+  This class is for the algorithms of Physics-guided Coverage Mapping
+  It inherits from the PostProcessor directly
+  ##TODO: Recast it once the new PostProcesso API gets in place
```
please update this info, since the new Postprocessor API is already in raven devel

huang714 replied:
Updated.

wangcj05 commented on framework/Metrics/metrics/MetricUtilities.py:
```diff
@@ -21,6 +21,7 @@
 import numpy as np
 import math
 import scipy
+from scipy import stats
```
why do you need this import? I do not see the place that you are using this function.

huang714 replied:
import removed

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+
+class PhysicsGuidedCoverageMapping(ValidationBase):
+  """
+    PhysicsGuidedCoverageMapping is a base class for validation problems
+    It represents the base class for most validation problems
```
Please update the docstring to reflect the PhysicsGuidedCoverageMapping class

huang714 replied:
Updated.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+        specifying input of cls.
+    """
+    specs = super(PhysicsGuidedCoverageMapping, cls).getInputSpecification()
+    #specs.addSub(metricInput)
```
remove this line?

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    super().__init__()
+    self.printTag = 'POSTPROCESSOR PhysicsGuidedCoverageMapping'
+    self.dynamicType = ['static','dynamic'] #  for now only static is available
```
if only 'static' available now, please remove 'dynamic'.

huang714 replied:
'dynamic' removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+    self.dynamicType = ['static','dynamic'] #  for now only static is available
+    self.acceptableMetrics = ["CDFAreaDifference", "PDFCommonArea", "STDReduction"] #  acceptable metrics
+    self.name = 'PhysicsGuidedCoverageMapping'
+    # self.pivotParameter = None
```
remove this line?

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+    pivotParameter = self.pivotParameter
+    names = [self.getDataSetName(inp[-1]) for inp in inputIn['Data']]
+    if len(inputIn['Data'][0][-1].indexes) and self.pivotParameter is None:
+      if 'dynamic' not in self.dynamicType: #self.model.dataType:
```
remove self.model.dataType?

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+        self.raiseAnError(IOError, "The validation algorithm '{}' is not a dynamic model but time-dependent data has been inputted in object {}".format(self._type, inputIn['Data'][0][-1].name))
+    evaluation ={k: np.atleast_1d(val) for k, val in  self._evaluate(dataDict, **{'dataobjectNames': names}).items()}
+
+    if pivotParameter:
+      if len(inputIn['Data'][0][-1]['time']) != len(list(evaluation.values())[0]):
+        self.raiseAnError(RuntimeError, "The pivotParameter value '{}' has size '{}' and validation output has size '{}'".format( len(dataSets[0][self.pivotParameter]), len(evaluation.values()[0])))
+      if pivotParameter not in evaluation:
+        evaluation[pivotParameter] = inputIn['Data'][0][-1]['time']
```
I'm confused here, Can the class handle dynamic situation?

huang714 replied:
I don't know... pivotParameter is not used in my case, I don't actually understand how to use this pivotParameter.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      ymin = m2.min()
+      ymax = m2.max()
+
+      binKDE = 200j
```
why binKDE is a complex number?

huang714 replied:
When using np.mgrid to generate the mesh grid, using the complex number will include the stop/max value. Is it better to use real numbers?

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      ymax = m2.max()
+
+      binKDE = 200j
+      X, Y = np.mgrid[xmin:xmax:binKDE, ymin:ymax:binKDE]
```
I think you need to use "()" for function call.

huang714 replied:
It is an object of some class instead of a function.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      knl = stats.gaussian_kde(vals)
+      Z = np.reshape(knl(psts).T, X.shape)
+
+      # Virtual Measurement
+      msrMean = 0.0
+      msrStd = 0.01*np.std(yAppStd)
+      msrNumSmpl = 1000
+      randomUtils.randomSeed(0)
+      yMsr = msrStd * randomUtils.randomNormal(size=msrNumSmpl) + msrMean
+      binMsr = msrNumSmpl//20
+      yMsrPdf, yMsrBin = np.histogram(yMsr, binMsr, range=(xmin, xmax), density=True)
```
Data for measurement should be provided through csv file, which can be loaded by raven, instead of hard-coded here.

huang714 replied:
I don't know how to add another input variable here, other methods in this validation gate only requires Features and Targets, no measurement information required.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      priStd = np.std(yAppStd)
+      stdReduct = 1.0-predStd/priStd
+
+      # Generate distribution by yAppPredPdfNorm
+      predBins = np.insert(Y[0, :]+0.5*(Y[0, 1]-Y[0, 0]), 0, Y[0, 0]-0.5*(Y[0, 1]-Y[0, 0]))
+      predHist = tuple((yAppPredPdfNorm, predBins))
+      predDist = stats.rv_histogram(predHist)
+      yAppPred = predDist.rvs(size=len(yAppStd), random_state=randomUtils.randomSeed(0, seedBoth=True))
```
It is not clear to me how this was calculated. First, I think the user should be able to provide the seed, second, when the seed changes, yAppPred will change, which will affect your results, is it what you want?

huang714 replied:
Modified to use variables calculated

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      priData = tuple((yAppStd.reshape(len(yAppStd),1), targData[1]))
+      postData = tuple((yAppPred.reshape(len(yAppStd),1), targData[1]))
+      for metric in self.metrics:
+        #name = "{}_{}_{}".format(feat.split("|")[-1], targ.split("|")[-1], metric.estimator.name)
```
remove this line?

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+
+    return outputDict
+
+  def _getDataFromDataDict(self, datasets, var, names=None):
```
this function should be moved to ValidationBase class.

huang714 replied:
Agree.

huang714 replied:
What should I do here?

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,218 @@
+      featData = self._getDataFromDataDict(datasets, feat, names)
+      targData = self._getDataFromDataDict(datasets, targ, names)
+
+      # Standardization
+      yExp = np.array(featData[0]).reshape(len(featData[0]))
+      yApp = np.array(targData[0]).reshape(len(targData[0]))
```
what is your data look like here? I guess the return from self._getDataFromDataDict will be a two dimension data, the first dimension is the number of realizations and the second dimension is the number of time steps. You are choosing the first row here, why is that? What is your assumption here? Do you assume the data is static? Can the method handle time-dependent responses?

huang714 replied:
The current implementation cannot handle time-dependent responses. In my case, values in the second dimension are all 1, so I used the first row. It only works for static data.

wangcj05 commented on tests/framework/AnalyticModels/tensor_poly2.py:
```diff
@@ -0,0 +1,31 @@
+#  This is a simple polynomial evaluation of all single-order combination polynomials
+#    For instance, xyz + xy + xz + yz + x + y + z + 1
+#
+import numpy as np
```
what is the different between this file and the existing tenso_poly file?

huang714 replied:
The output variable name is 'ans2' instead of 'ans' in tensor_poly file. I need to use output values of two different external models, whose name are both 'ans' in their model file, I don't know how to specify two variables with the same name in my .xml file...

huang714 replied:
I changed the ExternalModel to projectile.py

huang714 replied:
This file is not in use in current test.xml

wangcj05 reviewed (changes requested):
@huang714  Thanks for your contribution. I have some comments for your consideration. Please let me know if you have any question.

huang714 commented:
I have updated the pcm scripts to enable multiple responses of the Experiments/Features and fixed several issues raised in the comments. Would you please start another review process? @wangcj05 @Jimmy-INL Thank you!

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    ymin = m2.min()
+    ymax = m2.max()
+
+    binKDE = 200j
```
Move this to class __init__ method, or possibly let the user to decide the value

huang714 replied:
Moved to init.

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    # on range [ymsr.min(), ymsr.max()]
+    pdfAppPred = np.zeros(Y.shape[1])
+    intgrPdf = 0.0 # for normalization
+    for i in range(len(Y[0, :])):
+      pAppi = 0.0
+      for j in range(len(X[:, 0])):
+        pAppi += knl.evaluate([X[j, 0], Y[0, i]]) * np.diff(X[:, 0])[0] * pdfMsr[j]
+      pdfAppPred[i] = pAppi
+      intgrPdf += pdfAppPred[i]*(Y[0, 1]-Y[0, 0])
```
First, could you check if there is a better way to do the integration?
Second, could you check the equation? it seems to me you need to integrate over the cell (4 point in X, Y space).

huang714 replied:
Not in use any more, removed.

huang714 replied:
Changed to dot product.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    pdfAppPredNorm = pdfAppPred/intgrPdf
+
+    # Calculate Expectation (average value) of predicted application
+    predMean = 0.0
+    for i in range(len(Y[0, :])):
+      predMean += Y[0, i]*pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+    # Calculate Variance of predicted application
+    predVar = 0.0
+    for i in range(len(Y[0, :])):
+      predVar += (Y[0, i]-predMean)**2.0 * pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
```
I'm not clear about these equations, please try to prepare a document for these calculations.

huang714 replied:
Changed to a static model.

huang714 replied:
Used the expressions of expected value and variance from continuous PDF.

wangcj05 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    priStd = np.std(yAppStd)
+    stdReduct = 1.0-predStd/priStd
+
+    for metric in self.metrics:
+      name = "pri_post_{}".format(metric.estimator.name)
+      outputDict[name] = stdReduct
```
It seems to me metrics  are not used.

huang714 replied:
Removed.

wangcj05 commented on framework/Models/PostProcessors/Validations/Probabilistic.py:
```diff
@@ -1,145 +0,0 @@
-from ..ValidationBase import ValidationBase
-#Internal Modules End--------------------------------------------------------------------------------
-
-class Probabilistic(ValidationBase):
```
why this class got removed?

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm.xml:
```diff
@@ -0,0 +1,181 @@
+
+  <Models>
+    <ExternalModel ModuleToLoad="../../../AnalyticModels/projectile.py" name="exp" subType="">
+      <variables>x0,y0,v0,angle,timeOption,y,t</variables>
+    </ExternalModel>
+    <ExternalModel ModuleToLoad="../../../AnalyticModels/projectile.py" name="app" subType="">
+      <variables>x0,y0,v0,angle,timeOption,r</variables>
+    </ExternalModel>
+    <ExternalModel ModuleToLoad="../../../AnalyticModels/projectile.py" name="msr" subType="">
+      <variables>x0,y0,v0,angle,timeOption,y,t</variables>
+    </ExternalModel>
```
please use <inputs> and <outputs> instead of <variables>

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm.xml:
```diff
@@ -0,0 +1,181 @@
+      <Features>outputDataMC1|y</Features>
+      <Targets>outputDataMC2|r</Targets>
+      <Measurements>msrData|y</Measurements>
+      <Metric class="Metrics" type="Metric">std_reduct</Metric>
```
Metric is not used in the implementation

wangcj05 commented on tests/framework/AnalyticModels/tensor_poly2.py:
```diff
@@ -0,0 +1,31 @@
+#
+#  This is a simple polynomial evaluation of all single-order combination polynomials
+#    For instance, xyz + xy + xz + yz + x + y + z + 1
+#
```
Are you using this model in your test?

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm.xml:
```diff
@@ -0,0 +1,181 @@
+      <Input class="DataObjects" type="PointSet">inputPlaceHolder1</Input>
+      <Model class="Models" type="ExternalModel">app</Model>
+      <Sampler class="Samplers" type="MonteCarlo">MC_external</Sampler>
+      <Output class="DataObjects" type="PointSet">outputDataMC2</Output>
```
Why this model output is PointSet, but others are HistorySet?

wangcj05 reviewed (changes requested):
I have a couple of comments for your consideration

huang714 commented on framework/Metrics/metrics/STDReduction.py:
```diff
@@ -0,0 +1,51 @@
+from .MetricInterface import MetricInterface
+from Metrics.metrics import MetricUtilities
+
+class STDReduction(MetricInterface):
```
No, I don't need the STDReduction Metric now. But I don't know how I can define a metric in the input.xml if I don't have a Metric specified.

huang714 commented on framework/Models/PostProcessors/Factory.py:
```diff
@@ -37,7 +37,8 @@
 from .Validations import Probabilistic
-from .Validations import PPDSS
+from .Validations import PhysicsGuidedCoverageMapping
+#from .Validations import PPDSS
```
Added back. Previously I didn't have PPDSS in my version, had to remove this line to execute.

huang714 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+      @ Out, None
+    """
+    super()._handleInput(paramInput)
+    for child in paramInput.subparts:
+      if child.getName() == 'Measurements':
+        self.measurements = child.value    
```
'type' removed

huang714 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    """
+    names = kwargs.get('dataobjectNames')
+    outputDict = {}
+    for feat, targ, msr in zip(self.features, self.targets, self.measurements):
```
Modified.

huang714 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    """
+    names = kwargs.get('dataobjectNames')
+    outputDict = {}
+    for feat, targ, msr in zip(self.features, self.targets, self.measurements):
+      featData = self._getDataFromDataDict(datasets, feat, names)
+      targData = self._getDataFromDataDict(datasets, targ, names)
+      msrData = self._getDataFromDataDict(datasets, msr, names)
```
For loop is preserved since names from self.<x> is a list, while the function _getDataFromDataDict can only recognize string.

huang714 replied:
Back after updating.

huang714 commented on framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,202 @@
+    yExpRef = np.mean(yExp, axis=0)
+    yAppRef = np.mean(yApp)
+    # Standardization
+    yExpStd = (yExp-yExpRef)/yExpRef
+    yAppStd = (yApp-yAppRef)/yAppRef
+    if yExp.shape[1]!= yMsr.shape[1]:
+      self.raiseAnError(IOError, "Number of Measurements is not consistent with number of Experiments/Features.")
+    yMsrStd = (yMsr-yExpRef)/yExpRef
```
The reference value is usually given, PCM will map the uncertainty relative to the reference value. Will be documented.

huang714 commented:
Updated per your comments, changed to static responses, enabled multiple responses, and tested with linear model. Would you please take a look? @wangcj05 @Jimmy-INL Still a few issues remaining, will discuss tomorrow. Thank you!

huang714 commented:
I have updated what we discussed in yesterday's meeting. Would you please start a review of it? @wangcj05 @Jimmy-INL Thank you.

wangcj05 commented:
@huang714 Could you resolve the conflicting files in this PR?

huang714 commented:
> @huang714 Could you resolve the conflicting files in this PR?

It shows 'Already up to data' after 'git merge devel'. I couldn't find the three conflict files now.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,213 @@
+      X, Y = np.mgrid[xmin:xmax:self.binKDE, ymin:ymax:self.binKDE]
+      psts = np.vstack([X.ravel(), Y.ravel()])
+      vals = np.vstack([m1, m2])
+      # Measurement PDF over Exp range
+      pdfMsr = knlMsr(X[:, 0])
+
+      # Check whether the covavariance matrix is positive definite
+      if np.linalg.cond(vals)>=1/np.finfo(vals.dtype).eps:
+        pdfAppPred = knlMsr(Y[0, :])
+      # If not, introduce a
+      else:
+        knl = stats.gaussian_kde(vals)
+        # Joint PDF of Experiment and Application
+        Z = np.reshape(knl(psts).T, X.shape)
+        # yAppPred by integrating p(yexp, yapp)p(ymsr) over [yexp.min(), yexp.max()]
+        pdfAppPred = np.dot(Z, pdfMsr.reshape(pdfMsr.shape[0], 1))
+
+      # Normalized PDF of predicted application
+      pdfAppPredNorm = pdfAppPred.flatten()/pdfAppPred.sum()/np.diff(Y[0, :])[0]
+
+      # Calculate Expectation (average value) of predicted application
+      # by integrating xf(x), where f(x) is PDF of x
+      predMean = 0.0
+      for i in range(len(Y[0, :])):
+        predMean += Y[0, i]*pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+      # Calculate Variance of predicted application
+      # by integrating (x-mu_x)^2f(x), where f(x) is PDF of x
+      predVar = 0.0
+      for i in range(len(Y[0, :])):
+        predVar += (Y[0, i]-predMean)**2.0 * pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+      # Predicted standard deviation is square root of variance
+      predStd = np.sqrt(predVar)
+      # Prior standard deviation is the sample standard deviation
+      # Consider probability weights in the future
+      priStd = np.std(yAppStd)
+      # Uncertainty reduction fraction is 1.0-sigma_pred/sigma_pri
+      name = "pri_post_stdReduct_" + targ.split('|')[-1]
+      outputDict[name] = (1.0-predStd/priStd)
```
Please check the "for" loop here, I think several of the calculations can be moved to the outside of the for loop. For example, the calculation for yExpStd, yMsrStd. Please go through the for loops.

huang714 replied:
Data, reference, and standardization from Experimental side are moved outside the loop.

wangcj05 commented on tests/framework/AnalyticModels/linModel.py:
```diff
@@ -0,0 +1,46 @@
+  """
+  self.F1,self.F2,self.F3 = main(Input)
+
+def main(Input):
```
docstring for this method.

huang714 replied:
Modified.

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm.xml:
```diff
@@ -0,0 +1,151 @@
+    <name>framework/PostProcessors/Validation/test_validation_gate_pcm</name>
+    <author>dhuang</author>
+    <created>2021-07-21</created>
+    <classesTested>PostProcessors.Validation</classesTested>
```
update this info to reflect the PCM class

huang714 replied:
Updated.

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm.xml:
```diff
@@ -0,0 +1,151 @@
+    <classesTested>PostProcessors.Validation</classesTested>
+    <description>
+        This test is aimed to show how PCM works.
+        For simplicity, this test is using attenuate as experiment and tensor_ploy as application.
+        The output name convention is ...``feature name''\_``target name''\_``metric name''.
```
Update this comments to reflect the model and the outputs of PCM.

huang714 replied:
Updated.

wangcj05 reviewed (changes requested):
@huang714 I have a couple of comments for you to review.

huang714 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,213 @@
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from utils import InputData, InputTypes
```
Changed

huang714 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,213 @@
+      @ Out, outputDict, dict, dictionary containing the post-processed results
+    """
+    dataDict = {self.getDataSetName(data): data for _, _, data in inputIn['Data']}
+    names = [self.getDataSetName(inp[-1]) for inp in inputIn['Data']]
```
Modified, but haven't verified.

huang714 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -0,0 +1,213 @@
+
+  def _evaluate(self, datasets, **kwargs):
+    """
+      Main method to "do what you do".
```
Updated.

wangcj05 commented:
@huang714 white space issue, could you resolve it?
![image](https://user-images.githubusercontent.com/7321071/163004262-2712c998-2c2c-4155-b099-efa6c4180abd.png)

wangcj05 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,18 @@
+{
```
remove

huang714 replied:
removed

wangcj05 commented on .vscode/raven.code-workspace:
```diff
@@ -0,0 +1,8 @@
+{
```
remove

huang714 replied:
removed

wangcj05 commented on doc/user_manual/PostProcessors/PhysicsGuidedCoverageMapping.tex:
```diff
@@ -0,0 +1,32 @@
+\label{PhysicsGuidedCoverageMapping}
```
move to Validation.tex

huang714 replied:
Content moved to Validation.tex and the separate .tex file removed.

wangcj05 commented on doc/user_manual/postprocessor.tex:
```diff
@@ -113,6 +113,119 @@ \subsection{PostProcessor}
+  <Models>
+  ...
+<Simulation>
+\end{lstlisting}
```
remove all the changes here.

huang714 replied:
Removed

wangcj05 reviewed (changes requested):
I have some comments for your consideration.

huang714 commented:
I dealt with the comments on the doc file, however, I cannot compile the user manual to check whether the changes are displayed correctly. The problem seems to be a version.tex file under raven/doc/ missing (raven_user_manual.tex line 441). Please let me know if there is any other issue. Thanks! @wangcj05 @Jimmy-INL

joshua-cogliati-inl commented:
> I dealt with the comments on the doc file, however, I cannot compile the user manual to check whether the changes are displayed correctly. The problem seems to be a version.tex file under raven/doc/ missing (raven_user_manual.tex line 441). Please let me know if there is any other issue. Thanks! @wangcj05 @Jimmy-INL

FYI: to see the problems with generating the documents, run:
```./make_docs.sh --verbose```

wangcj05 commented:
Hi @huang714, fyi, the following test failed on most of the Linux machines: 

![image](https://user-images.githubusercontent.com/7321071/169381106-3d89b0d9-e85e-45b9-b968-0c4dbf5dc28f.png)

huang714 commented:
> Hi @huang714, fyi, the following test failed on most of the Linux machines:
> 
> FAILED: Diff tests/framework/PostProcessors/Validation/test_validation_gate_pcm
> 
> ![image](https://user-images.githubusercontent.com/7321071/169381106-3d89b0d9-e85e-45b9-b968-0c4dbf5dc28f.png)

This is because I used a system error related condition to determine whether this is singular matrix. The new condition is changed to condition number greater than 100. Let me know if there is a better way. Thank you!

moosebuild commented:
All [jobs](https://civet.inl.gov/event/92921/) on 1ec4845 : invalidated by @wangcj05

retesting

moosebuild commented:
All [jobs](https://civet.inl.gov/event/92921/) on 1ec4845 : invalidated by @wangcj05

retesting

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1090151/) on 1ec4845 : invalidated by @wangcj05

fetch issue

wangcj05 commented on doc/user_manual/PostProcessors/Validation.tex:
```diff
@@ -148,3 +149,42 @@ \subsubsection{Validation PostProcessors}
+    number of measurements should be equal to the number of features and in the same order as the features listed in \xmlNode{Features}.
+\end{itemize}
+
+The output of PCM is comma separated list of strings in the format of "pri\_post\_stdReduct\_[targetName]", 
```
please change "pri\_post\_stdReduct\_[targetName]" to ``pri\_post\_stdReduct\_[targetName]''

huang714 replied:
Modified.

wangcj05 commented on ravenframework/Models/PostProcessors/Factory.py:
```diff
@@ -38,6 +38,8 @@
 from .Validations import Probabilistic
 from .Validations import PPDSS
+from .Validations import PhysicsGuidedCoverageMapping
+#from .Validations import PPDSS
```
remove this line

huang714 replied:
Removed.

wangcj05 reviewed (commented):
some minor comments, @huang714 could you fix my comments?

wangcj05 reviewed (approved):
PR looks good.

wangcj05 commented on doc/user_manual/PostProcessors/Validation.tex:
```diff
@@ -148,3 +149,42 @@ \subsubsection{Validation PostProcessors}
+    number of measurements should be equal to the number of features and in the same order as the features listed in \xmlNode{Features}.
+\end{itemize}
+
+The output of PCM is comma separated list of strings in the format of ``pri_post_stdReduct_[targetName]", 
```
use "\_" for "_" otherwise the manual can not be built. @huang714

huang714 replied:
I changed to "\textunderscore" to fix this issue and it showed correctly in the built manual. Using any "_" symbol cannot be built from my side. I have also modified multiple places of PCM doc. Thx

wangcj05 reviewed (approved):
changes are good.

moosebuild commented:
Job [Test qsubs sawtooth](https://civet.inl.gov/job/1143805/) on 060341b : invalidated by @wangcj05

FAILED: Diff tests/cluster_tests/test_mpiqsub_parameters Diff tests/cluster_tests/test_mpiqsub_nosplit

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1143806/) on 060341b : invalidated by @wangcj05

TimeoutError: Timed out while waiting for node to startup.

wangcj05 commented:
checklist is good, and tests passed.
